plugin/storage/es/spanstore/writer.go

```diff
@@ -42,6 +44,63 @@ type SpanWriter struct {
 	serviceWriter    serviceWriter
 	spanConverter    dbmodel.FromDomain
 	spanServiceIndex spanAndServiceIndexFn
+	indexCache       sync.Map
+}
+
+func (s *SpanWriter) ensureIndex(ctx context.Context, indexName string) error {
+	if _, exists := s.indexCache.Load(indexName); exists {
+		return nil
+	}
+
+	_, loaded := s.indexCache.LoadOrStore(indexName, struct{}{})
+	if loaded {
+		return nil
+	}
+
+	exists, err := s.client().IndexExists(indexName).Do(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to check index existence: %w", err)
+	}
+
+	if !exists {
+		s.logger.Info("Creating index", zap.String("index", indexName))
+
+		// Set specific settings for the test environment
+		body := `{
+            "settings": {
+                "number_of_shards": 1,
+                "number_of_replicas": 0,
+                "index.write.wait_for_active_shards": 1
+            }
+        }`
+
+		_, err = s.client().CreateIndex(indexName).Body(body).Do(ctx)
```

Comment:
the writer should not be responsible for creating indices. The indices are created by ES automatically from templates on writes.

The templates are created by the factory: https://github.com/jaegertracing/jaeger/blob/61ad598c76f85b4bd8292048c80d714185977650/plugin/storage/es/factory.go#L279

Reply:
@yurishkuro 

I see,  now i understand the core issue better:

1. it's  about template creation timing vs span writing.
2. Indices are automatically created by ES based on templates.
3. The race condition occurs in factory.go where templates are created and then spans are written before templates are fully ready?

Would this be a better approach?

1. Remove all index management code from SpanWriter
2. Add template creation synchronization in factory.go's createSpanWriter,

something like, passing context in createSpanWriter and in Create Templates function we can handle the context.. I am taling about these 2 functions, if thinking in correct direction shall i proceed and test?

<div style="display: flex; gap: 20px;">
<img src="https://github.com/user-attachments/assets/e179c929-c259-44d1-ae87-9d1a4baf07e7" width="45%" />
<img src="https://github.com/user-attachments/assets/6211ccfa-b664-4c12-a302-e21de5b809d5" width="45%" />
</div>

Reply:
Visualizing like this (What do you think, shall i proceed with this?)

Before:
Start -> Create Writer -> Start Template Creation -> Return Writer -> Write Spans (fails since template still creating)
                                     [Templates Still Creating...]

After:
Start-> Create Writer -> Create Templates -> Retry if needed -> Templates Ready -> Return Writer -> Write Spans

Reply:
Yes, we want the sequence in the After part, we just need to enforce that the templates finished creating before returning the writer.

Reply:
made a new PR, #6229